### PR TITLE
Norm preserving initialization for WeightNorm

### PR DESCRIPTION
**Pull request by victorcampos7**
This PR implements the initialization for WeightNorm described in [Arpit et al. (2019)](https://arxiv.org/abs/1906.02341)

The paper shows improvements for this initialization scheme over the data-dependent initialization (which is not currently implemented in PyTorch) and the current default PyTorch initialization.

### CODE REVIEW

File: torch/nn/utils/weight_norm.py [hunk 1]
```diff
@@ -34,8 +39,15 @@ def apply(module, name, dim):
         # remove w from parameter list
         del module._parameters[name]
 
+        # initial value for g
+        if init == 'default':
+            g_init = norm_except_dim(weight, 2, dim)
+        elif init == 'norm_preserving':
+            fan_in, fan_out = _calculate_fan_in_and_fan_out(weight)
+            g_init = ones_like(norm_except_dim(weight, 2, dim)) * math.sqrt(gamma * fan_in / fan_out)
```
Comment (apaszke) [inline torch/nn/utils/weight_norm.py]: `full_like` will make this much faster.

**Review by apaszke — COMMENTED** [review]
It's a new paper, has this been used anywhere else? We're generally a bit careful about including new techniques in the core library because many of them quickly get out of fashion, and would only make our code unnecessarily complex for backward compatibility reasons.

File: torch/nn/utils/weight_norm.py (hunk 1 above)
Comment (victorcampos7) [reply torch/nn/utils/weight_norm.py]: Thanks for the suggestion, I'll update the code to use `full_like`.

**Comment by victorcampos7** [comment]
It's a new paper, and I don't think this init strategy has been used anywhere else yet. The motivation for submitting this PR is that the current initialization for WeightNorm does not seem to be based on any existing publication (please let me know if it is) and can result in vanishing/exploding gradients. The only way to modify the initialization for `weight_g`with the current implementation is to initialize it to the default value and override it later.

I understand your point about not including new techniques until they're proven useful. We can always wait and reconsider adding this feature in the future if it becomes useful to the community.

**Comment by apaszke** [comment]
cc @SsnL (I think you have implemented large parts of WeightNorm?)

**Review by ssnl — COMMENTED** [review]
I wrote most of SpectralNorm, not WeightNorm. But I think this is great to have. I don't see any correctness test wrt the norm preserving behavior (e.g., asserting that relevant norm is constant or something). Could you add one?

**Comment by victorcampos7** [comment]
Norm is preserved in expectation, I'm not sure that we can add a correctness test for this. I can add other tests, e.g. checking that the initial value of `weight_g`is correct. Would that be enough?

**Comment by victorcampos7** [comment]
I added a correctness test for the norm preserving behavior. Given that norm preservation is achieved in expectation, the test checks that the ratio of output/input norms is close to 1 in a wide layer with 5000 units.

**Comment by soumith** [comment]
@colesbury do you remember where you added the weight_norm initialization from?
Also requesting a review from you.

**Comment by colesbury** [comment]
> colesbury do you remember where you added the weight_norm initialization from?

It's a reparameterization. In general, there's only one choice of `g` that preserves the original function.

**Comment by victorcampos7** [comment]
Hi @colesbury, there's a review that has been pending for 29 days. I don't know if it usually takes this long, so this is just a gentle reminder in case it has been forgotten.

**Comment by soumith** [comment]
@victorcampos7 I think we've been trying to make the determination on whether to actually accept this.

As @colesbury pointed out, you are actually changing the output function, and it's no longer a re-parameterization, so it opens a whole other set of weirdness -- it preserves the norm but it's not computing the same function anymore.

**Comment by victorcampos7** [comment]
@soumith you're right -- with this init scheme, we are indeed changing the output function. If you want PyTorch's WeightNorm implementation to keep the original function, then the norm preserving init doesn't really fit here.

However, as fair as I understand, keeping the original function is not the main goal of WeightNorm. Section 3 in [the original paper](https://arxiv.org/abs/1602.07868) discusses initialization schemes for WeightNorm with the goal of making optimization easier (which the norm preserving init does). Indeed, the [reference Theano/TensorFlow implementation by OpenAI](https://github.com/openai/weightnorm) uses data-dependent initialization -- which modifies the original function.
